pkg/api/service_test.go

```diff
@@ -112,13 +124,247 @@ func (s *ServiceTestSuite) TestInstallShouldReturnResultOnSuccess() {
+	t := s.T()
+	assert.NoError(t, err)
+	require.NotNil(t, res)
+	assert.Equal(t, res.Status, "deployed")
+	s.chartloader.AssertExpectations(t)
+	s.upgrader.AssertExpectations(t)
+}
+
+func (s *ServiceTestSuite) TestUpgradeValidateFailShouldResultFailure() {
+	chartName := "./some/local-chart"
+	cfg := api.ReleaseConfig{
+		Name:      "some-component",
+		Namespace: "hermes",
+		ChartName: chartName,
+	}
+	var vals api.ChartValues
+
+	res, err := s.svc.Upgrade(s.ctx, cfg, vals)
+
+	t := s.T()
+	assert.Nil(t, res)
+	assert.EqualError(t, err, "error request validation: cannot refer local chart")
+	s.chartloader.AssertNotCalled(t, "LocateChart")
+	s.upgrader.AssertNotCalled(t, "SetConfig")
+	s.upgrader.AssertNotCalled(t, "Run")
+}
+
+func (s *ServiceTestSuite) TestUpgradeShouldReturnErrorOnInvalidChart() {
+	chartName := "stable/invalid-chart"
+	cfg := api.ReleaseConfig{
+		Name:      "some-component",
+		Namespace: "hermes",
+		ChartName: chartName,
+	}
+	var vals api.ChartValues
+	s.chartloader.On("LocateChart", chartName, s.settings).Return("", errors.New("Unable to find chart"))
+
+	res, err := s.svc.Upgrade(s.ctx, cfg, vals)
+
+	t := s.T()
+	assert.Nil(t, res)
+	assert.EqualError(t, err, "error in locating chart: Unable to find chart")
+	s.chartloader.AssertExpectations(t)
+	s.upgrader.AssertNotCalled(t, "SetConfig")
+	s.upgrader.AssertNotCalled(t, "Run")
+}
+
+func (s *ServiceTestSuite) TestListShouldReturnErrorOnFailureOfListRun() {
+	s.lister.On("SetState", action.ListDeployed)
+	s.lister.On("SetStateMask")
+
+	var releases []*release.Release
+
+	s.lister.On("Run").Return(releases, errors.New("cluster issue"))
+
+	releaseStatus := "deployed"
+	res, err := s.svc.List(releaseStatus)
+
+	t := s.T()
+	assert.Error(t, err, "cluster issue")
+	assert.Nil(t, res)
+
+	s.lister.AssertExpectations(t)
+}
+
+func (s *ServiceTestSuite) TestListShouldReturnAllReleasesIfNoFilterIsPassed() {
+	layout := "2006-01-02T15:04:05.000Z"
+	str := "2014-11-12T11:45:26.371Z"
+	timeFromStr, err := time.Parse(layout, str)
+
+	s.lister.On("SetState", action.ListAll)
+	s.lister.On("SetStateMask")
+
+	var releases []*release.Release
+	releases = append(releases,
+		&release.Release{Name: "test-release",
+			Namespace: "test-namespace",
+			Info:      &release.Info{Status: release.StatusDeployed, LastDeployed: timeFromStr},
+			Version:   1,
+			Chart:     &chart.Chart{Metadata: &chart.Metadata{Name: "test-release", Version: "0.1", AppVersion: "0.1"}},
+		})
+
+	s.lister.On("Run").Return(releases, nil)
+
+	releaseStatus := ""
+	res, err := s.svc.List(releaseStatus)
+
+	t := s.T()
+	assert.NoError(t, err)
+	require.NotNil(t, res)
+
+	var response []api.Release
+	response = append(response, api.Release{Name: "test-release",
+		Namespace:  "test-namespace",
+		Revision:   1,
+		Updated:    timeFromStr,
+		Status:     release.StatusDeployed,
+		Chart:      "test-release-0.1",
+		AppVersion: "0.1",
+	})
+
+	assert.Equal(t, 1, len(res))
```

Comment:
`require.Equal` use to verify length of slice before comparison, in case of failures the test flows through and panics